src/Ball.java

```diff
@@ -83,14 +120,102 @@ public void draw() {
      * Moves the ball so that the next time it draws it will be in a different place
      */
     public void move() {
-        x = x + xSpeed;
-        y = y + ySpeed;
-        if (x > s.width - radius || x < radius) {
-            xSpeed = -xSpeed;
-        }
-        if (y > s.height - radius || y < radius) {
-            ySpeed = -ySpeed;
+        if (!frozen) {
+            x = x + xSpeed;
+            y = y + ySpeed;
+
+            if (x >= s.width - radius - 5 || x <= radius + 5) {
+                xSpeed = -xSpeed;
+            }
+            if (y >= s.height - radius - 5 || y <= radius + 5) {
+                ySpeed = -ySpeed;
+            }
         }
     }
 
-}
\ No newline at end of file
+    /**
+     * Returns true if the ball is colliding with another ball
+    */
+    public boolean collidesWith(Ball other) {
+        float dx = x - other.x;
+        float dy = y - other.y;
+
+        // pythagoeran theorem
+        // tried to use processing's dist() function outside of the Sketch class and was being weird
+        float distance = (float)Math.sqrt((dx * dx) + (dy * dy));
+
+        // return if the distance between the two balls is less than the sum of their radiuses
+        return distance < (radius + other.radius);
+    }
+
+    /**
+     * Returns true if the ball is colliding with a bubble
+    */
+    public boolean collidesWithBubble(Bubble other) {
+        float dx = x - other.x;
+        float dy = y - other.y;
```

Comment:
oh no, did you make x and y public in Bubble to make this work? Bad! use the accessors.